app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go

```diff
@@ -46,32 +49,115 @@ const (
+	// Create health checks to be used by the backend service.
+	healthChecks, hcErr := l.hcs.EnsureHealthCheck(l.lbName, ports)
+	if hcErr != nil {
+		// Keep aggregating errors and return all at the end, rather than giving up on the first error.
+		err = multierror.Append(err, hcErr)
+	}
+	igs, namedPorts, err := l.getIGsAndNamedPorts(ing)
+	// Cant really create any backend service without named ports. No point continuing.
+	if err != nil {
+		return err
+	}
+	// Create backend service. This should always be called after the health check since backend service needs to point to the health check.
+	if err := l.bss.EnsureBackendService(l.lbName, ports, healthChecks, namedPorts, igs); err != nil {
 		return err
 	}
 	return nil
 }
 
+// Returns links to all instance groups and named ports on any one of them.
+// Note that it picks an instance group at random and returns the named ports for that instance group, assuming that the named ports are same on all instance groups.
+// Also note that this returns all named ports on the instance group and not just the ones relevant to the given ingress.
+func (l *LoadBalancerSyncer) getIGsAndNamedPorts(ing *v1beta1.Ingress) ([]string, backendservice.NamedPortsMap, error) {
+	// Keep trying until ingress gets the instance group annotation.
+	for try := true; try; time.Sleep(1 * time.Second) {
+		// Fetch the ingress from a cluster.
+		ing, err := l.getIng(ing.Name, ing.Namespace)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error %s in fetching ingress", err)
+		}
+		if ing.Annotations == nil || ing.Annotations[instanceGroupsAnnotationKey] == "" {
+			// keep trying
+			fmt.Println("Waiting for ingress to get", instanceGroupsAnnotationKey, "annotation.....")
+			continue
+		}
+		annotationValue := ing.Annotations[instanceGroupsAnnotationKey]
+		glog.V(3).Infof("Found instance groups annotation value: %s", annotationValue)
+		// Get the instance group name.
+		type InstanceGroupsAnnotationValue struct {
+			Name string
+			Zone string
+		}
+		var values []InstanceGroupsAnnotationValue
+		if err := json.Unmarshal([]byte(annotationValue), &values); err != nil {
+			return nil, nil, err
+		}
+		if len(values) == 0 {
+			// keep trying
+			fmt.Printf("Waiting for ingress to get", instanceGroupsAnnotationKey, "annotation.....")
+			continue
+		}
+		// Compute link to all instance groups.
+		var igs []string
+		for _, ig := range values {
+			igs = append(igs, fmt.Sprintf("%s/instanceGroups/%s", ig.Zone, ig.Name))
+		}
+		// Compute named ports from the first instance group.
+		zone := getZoneFromURL(values[0].Zone)
+		ig, err := l.igp.GetInstanceGroup(values[0].Name, zone)
+		if err != nil {
+			return nil, nil, err
+		}
+		fmt.Printf("Fetched instance group: %s/%s, got named ports: %#v", zone, values[0].Name, ig.NamedPorts)
+		namedPorts := backendservice.NamedPortsMap{}
+		for _, np := range ig.NamedPorts {
+			namedPorts[np.Port] = np
+		}
+		return igs, namedPorts, nil
+	}
+	return nil, nil, nil
```

Comment:
this looks unreachable. does Go require you to have that?

Reply:
yes :)